SOCKS5 proxy support for the MQTT client (TCP)

tgorochowik opened the pull request:
The use-case is to be able to connect to a MQTT broker (cloud) from a corporate network. To do that a proxy is needed.

For test purposes though, the SOCKS5 proxy server is opened at the same ip as the MQTT broker.

The following assumptions are made:
```
10.0.0.116 - IP address of the proxy server and the mosquito broker
1883       - MQTT port
1080       - SOCKS5 proxy port
```

Sample setup:
```
$ sudo mosquitto -v -p 1883
```
And:
```
$ ssh -N -D 10.0.0.116:1080 10.0.0.116
```

With the default configuration of the `mqtt_publisher` sample - everything should work as earlier. After enabling `CONFIG_SOCKS` in menuconfig, the `mqtt_publisher` should attempt to connect to the broker through a proxy server.

The SOCKS5 implementation here is very limited. Basically only the features needed for the single use-case are implemented.

Thanks to the fact that no other strict SOCKS5 messages will be exchanged in this scenario, the original socks5 socket can be used by the mqtt lib after initializing the connection - this allows us to modify only the connect part of the mqtt lib and leave the rest untouched.

tgorochowik committed:
net: Add initial SOCKS5 support

This adds some very basic SOCKS5 proxy client support.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: mqtt: Add SOCKS5 proxy support

This commits adds a new MQTT transport. The purpose is to be able to
connect to a MQTT broker through a SOCKS5 proxy.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
samples: net: mqtt_publisher: Add proxy support

If appropriate configuration is enabled, assume (with the default
configuration) that the proxy server uses the same address as the broker
and the default 1080 port.

The default settings can be changed in the configuration header.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tbursztyka commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,157 @@
+	int fd;
+
+	fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+
```
you can remove that empty line

tbursztyka commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	ret = connect(fd, proxy, sizeof(struct sockaddr_in));
+
```
same here etc...

tbursztyka commented on subsys/net/lib/mqtt/mqtt_transport_socket_tcp.c:
```diff
@@ -29,26 +32,44 @@ int mqtt_client_tcp_connect(struct mqtt_client *client)
+			return -errno;
+		}
 	}
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```
Could you use if (IS_ENABLED () && !IS_ENABLED()) ?

tgorochowik replied:
@tbursztyka Thanks! Could you elaborate (or maybe point me to the right place in the doc) on when should `#if defined(CONFIG_*)` and `#if IS_ENABLED(CONFIG_*)` be used? I am trying to fix the PR based on the input from you and others, and this confuses me the most, as `if defined(...)` seems to be used throughout the current code (in headers, lib code, and the sample app code too).

tbursztyka replied:
Well, we try to use IS_ENABLED as much as possible that's all. It does not mean we should never use #if defined(), there are places where you have to anyway.

tbursztyka commented on subsys/net/lib/mqtt/mqtt_transport_socket_tcp.c:
```diff
@@ -29,26 +32,44 @@ int mqtt_client_tcp_connect(struct mqtt_client *client)
-		(void)close(client->transport.tcp.sock);
-		return -errno;
+		client->transport.tcp.sock =
+			socks5_client_tcp_connect(proxy, broker);
```
you would need to make dummy inline in the socks header file, if CONFIG_SOCKS isn't enabled

tbursztyka commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -248,6 +251,13 @@ static void broker_init(void)
 	broker4->sin_family = AF_INET;
 	broker4->sin_port = htons(SERVER_PORT);
 	inet_pton(AF_INET, SERVER_ADDR, &broker4->sin_addr);
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```
can you use IS_ENABLED() etc...

Than will also make below proxy variable declared at the beginning of a code block, which would be proper unlike what it is right now.

tbursztyka commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -259,6 +269,9 @@ static void client_init(struct mqtt_client *client)
 
 	/* MQTT client configuration */
 	client->broker = &broker;
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```
same here

pfalcon commented on include/net/socks.h:
```diff
@@ -0,0 +1,14 @@
+
+#include <net/socket.h>
+
+int socks5_client_tcp_connect(const struct sockaddr*, const struct sockaddr*);
```
We require parameter names, and even docstrings for functions.

pfalcon commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,157 @@
+
+#include "socks_internal.h"
+
+/**@brief Connects to destination through a SOCKS5 proxy server.
```
So, this docstring should go to header. My understanding that API docs are generated from headers, or at least that's where we have docstrings in the rest of Zephyr.

pfalcon commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,157 @@
+	ret = send(fd, &cmd_req, size, 0);
+
+	if (ret != size) {
+		LOG_ERR("Could not send CONNECT command");
```
That won't work, sorry. It's not error to have only partial data to be sent. You need to retry with the rest. (Or maybe we need to implement MSG_WAITALL after all for convenience, but looking thru man, it's defined only for recv(), not for send()! ).

pfalcon replied:
> Or maybe we need to implement MSG_WAITALL after all for convenience, but looking thru man, it's defined only for recv(), not for send()!

This is now ticketed as https://github.com/zephyrproject-rtos/zephyr/issues/13091

tgorochowik replied:
I've added the helper functions based on how it was done in other places, however the current implementation is fully blocking. I guess that is ok in this case?

pfalcon replied:
> however the current implementation is fully blocking.

My response would be: "How else could it be? Writing non-blocking applications is rather complicated." But, you want to use it for mqtt lib, which went thru the trouble of implementing it in non-blocking way. Please work with @rlubos to ensure that it all will work as it should. I would think it should: you perform initial handshake in a blocking manner, then mqtt lib will just switch sock5 socket into non-blocking mode.

rlubos reviewed (changes requested):
I think that we should rather specify a new underlying transport for MQTT rather than slicing the existing one with ifdefs.

pfalcon commented:
> we should rather specify a new underlying transport for MQTT rather than slicing the existing one with ifdefs.

+1

> If we define `MQTT_TRANSPORT_SOCKS5` this field could be specific for that transport, again no need for ifdefs.

Note that absolutely any TCP protocol can work over SOCKS5. Will we add SOCKS5 option for each of them? There's nothing wrong also for any TCP protocol to work over Websocket. Websocket itself can work over SOCKS5. Sprinkle in TLS. Will we add options for each possible combination?

That's why I'm talking about composability in the other ticket. A user should have a freedom to compose any protocol pipeline they want. MQTT over Websocket over TLS over SOCKS5? Easy. And if they want HTTP over MQTT, it's not the protocol lib which should prevent them doing that.

tgorochowik commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -259,6 +269,9 @@ static void client_init(struct mqtt_client *client)
 
 	/* MQTT client configuration */
 	client->broker = &broker;
+#if defined(CONFIG_SOCKS) && !defined(CONFIG_MQTT_LIB_TLS)
```
@rlubos Thank you for all your comments.

It cannot really be done with runtime checks here, because some of the fields in the header are already `ifdef`'d - so we have to either include them at all times, or use the compile-time macros here too.
See:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/include/net/mqtt.h#L372-L383

Would leaving `#if defined(...)` here be ok? Or what do you suggest?

rlubos replied:
Ah I see the problem...

@tbursztyka How de we approach such issue, when some struct fields are conditionally included? Shall we remove `#if defined(...)` from the struct definition, or stick to `#if defined(...)` in code?

tgorochowik replied:
@rlubos 
There is also an issue with this:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/include/net/mqtt.h#L341-L355

If we add another transport, and explicitly assign it the next number (`0x02`), and then add the new definition here:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/subsys/net/lib/mqtt/mqtt_transport.c#L33-L48

And then say compile it with the new transport (`0x02`) enabled but the transport in between disabled (`CONFIG_MQTT_LIB_TLS` - `0x01`), then this will fail:
https://github.com/zephyrproject-rtos/zephyr/blob/bb50d75fc4ff146b38a25117b76f9bc720776bbb/subsys/net/lib/mqtt/mqtt_transport.c#L52

Because the enum value will no longer be equal to the array index.

So we have to either stop assigning the numbers in the enum explicitly, or handle the `#if..` differently here too.

rlubos replied:
Looks like getting rid `#if defined(CONFIG_MQTT_LIB_TLS)` is quite cumbersome in this case. Then I think we could stick to `#if defined(...)` nomenclature where needed, just as @tbursztyka commented earlier. 

Getting rid of explicit enum numeration sound good to me - currently it's kinda messy anyway (like fixed number of transports even though it's actually configurable).

tgorochowik committed:
net: mqtt: Remove explicit numeration of transport enum

The transport number is used to pick elements of an array which is
initialized using #if defined macros. Having explicit numeration in this
enum leads to situations where the array index is different than the
enum value.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik commented:
Thank you all for all the comments.

Since there have been quite a lot of changes then I guess a full re-review would be nice.

Things that still surly do require attention:
The _ifdeffing_ - most of it is removed because a new transport was added, however there are still places where it is used - and still `if defined` is used instead of `if IS_ENABLED` - in some cases the latter could not have been used, in some just because this is what was used in similar cases in the file - I am open to changing this, but I still think this should be at least consistent

jukkar commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,176 @@
+			+ sizeof(struct socks5_ipv4_addr);
+	} else if (proxy->sa_family == AF_INET6) {
+		const struct sockaddr_in6 *d6 =
+			(struct sockaddr_in6 *) destination;
```
No need to put extra space after a cast. Can you check and fix this everywhere in this PR

tgorochowik replied:
Right, thanks, I hope I didn't miss any!

jukkar commented on subsys/net/lib/socks/socks_internal.h:
```diff
@@ -0,0 +1,91 @@
+struct socks5_command_response {
+	struct socks5_command_response_common r;
+	union {
+		struct socks5_ipv4_addr ipv4_addr;
```
Can't we use "struct sockaddr" here?

pfalcon commented on subsys/net/lib/socks/socks.c:
```diff
@@ -0,0 +1,176 @@
+	while (offset < len) {
+		ret = send(fd, data + offset, len - offset, 0);
+		if (ret < 0) {
+			return -errno;
```
Why return errno here? Just return ret, caller can look up actual error in errno itself. That's standard POSIX protocol.

rlubos reviewed (approved):
One nit spotted worth fixing, otherwise I think it's good.

jukkar commented on subsys/net/lib/socks/socks_internal.h:
```diff
@@ -0,0 +1,91 @@
+	u8_t method;
+};
+
+struct socks5_ipv4_addr {
```
The socks5_ipv4_addr and socks5_ipv6_addr can be removed as "struct sockaddr" has both IP address and port number information in it.

tgorochowik replied:
But `struct sockaddr` also has other fields, and this particular struct is used directly as part of a buffer that is being sent/received.

Is there a way to work-around this without this extra struct?

tgorochowik replied:
That also makes me think whether `__packed` should not be added to the struct definitions.

pfalcon replied:
> and this particular struct is used directly as part of a buffer that is being sent/received.

> That also makes me think whether `__packed` should not be added to the struct definitions.

It it's sent/received over network, you definitely should.

tgorochowik replied:
Ok, just pushed an update with the attribute added to all the structs used as network buffers.

jukkar replied:
> But `struct sockaddr` also has other fields, and this particular struct is used directly as part of a buffer that is being sent/received.

Hmm, if this is only used here and if you think making the struct as small as possible here is important, then we can probably do it like this (at least for now).

jukkar replied:
> Hmm, if this is only used here and if you think making the struct as small as possible here is important, then we can probably do it like this (at least for now).

Never mind my comment, I did not realize that the struct was used to cast data from network. Then of course we need to do it like you suggested.

pfalcon reviewed (approved):
Thanks for refactors.

jukkar reviewed (approved):
LGTM

tbursztyka commented on subsys/net/lib/mqtt/mqtt_transport.c:
```diff
@@ -29,6 +29,11 @@ extern int mqtt_client_tls_read(struct mqtt_client *client, u8_t *data,
 extern int mqtt_client_tls_disconnect(struct mqtt_client *client);
 #endif /* CONFIG_MQTT_LIB_TLS */
 
+#if IS_ENABLED(CONFIG_MQTT_LIB_SOCKS)
```
IS_ENABLED is to be used only with code (at least that's the only usage we have), besides that: use #ifdef or #if defined()

Sorry if my comment about using IS_ENABLED was unclear.

tgorochowik replied:
ok, makes sense, thanks!

tgorochowik replied:
Fixed now.

tbursztyka commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -242,12 +245,28 @@ static void broker_init(void)
 	broker6->sin6_port = htons(SERVER_PORT);
 	inet_pton(AF_INET6, SERVER_ADDR, &broker6->sin6_addr);
+
+#if IS_ENABLED(CONFIG_MQTT_LIB_SOCKS)
```
So here is would be:

if (IS_ENABLED(CONFIG_MQTT_LIB_SOCKS)) {
...
}

tgorochowik replied:
But that won't work that easily. It uses things that are available only if the config is actually defined. So it either have to stay like this or we would have to include lots of stuff when it is not needed. That includes some defines from the configuration header, but in the later cases also fields of a struct that don't exist unless this config is enabled.

Besides that, a few lines above a similar check is made (CONFIG_NET_IPV6) and it is a compile time check too. Would the approach you're suggesting not make it too inconsistent?

tbursztyka commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -242,12 +245,28 @@ static void broker_init(void)
 	broker4->sin_port = htons(SERVER_PORT);
 	inet_pton(AF_INET, SERVER_ADDR, &broker4->sin_addr);
+
+#if IS_ENABLED(CONFIG_MQTT_LIB_SOCKS)
```
same here
